dali/kernels/imgproc/structure/connected_components.h

```diff
@@ -192,28 +264,80 @@ auto FullTensorSlice(const TensorView<StorageCPU, T, ndim> &tensor) {
  */
-template <typename OutLabel>
-int64_t CompactLabels(OutLabel *labels, int64_t volume, OutLabel bg_label = 0) {
+template <typename OutLabel, typename ExecutionEngine>
+int64_t CompactLabels(OutLabel *labels,
+                      int64_t volume,
+                      ExecutionEngine &engine,
+                      OutLabel bg_label = 0) {
   constexpr OutLabel old_bg_label = static_cast<OutLabel>(-1);
   OutLabel curr_label = 1;
-  std::set<OutLabel> label_set;
   std::unordered_map<OutLabel, OutLabel> label_map;
   disjoint_set<OutLabel, OutLabel> ds;
 
+  const int64_t chunk_size = 16<<10;
+  int num_chunks = std::min<int>(div_ceil(volume, chunk_size), engine.NumThreads());
+
   // Optimized label compaction algorithm:
   //
   // Labels are scanned and only when the label value changes are we invoking ds.find.
   // In the second pass, we only look up the map when the value changes, but this time we
   // can't call ds.find, because we're overwriting the labels array and the disjoint set
   // structure is effectively corrupted.
 
-  OutLabel prev = old_bg_label;
-  for (int64_t i = 0; i < volume; i++) {
-    if (labels[i] != old_bg_label && labels[i] != prev) {
-      // look up `ds` only when the value changes - this saves a lot of lookups
-      label_set.insert(ds.find(labels, i));
-      prev = labels[i];
+  std::set<OutLabel> label_set;
+
+  if (num_chunks == 1) {
+    std::unordered_set<OutLabel> tmp_set;
+    OutLabel prev = old_bg_label;
+    OutLabel remapped = old_bg_label;
+    for (int64_t i = 0; i < volume; i++) {
+      if (labels[i] != old_bg_label) {
+        if (labels[i] != prev) {
+          prev = labels[i];
+          // look up `ds` only when the value changes - this saves a lot of lookups
+          remapped = ds.find(labels, i);
+          tmp_set.insert(remapped);
+        } else {
+          labels[i] = remapped;
+        }
+      }
+    }
+    for (auto l : tmp_set)
+      label_set.insert(l);
+  } else {
+    SmallVector<std::unordered_set<OutLabel>, 16> tmp_sets;
+    tmp_sets.resize(engine.NumThreads());
+    spinlock lock;
+    for (int chunk = 0; chunk < num_chunks; chunk++) {
+      int64_t chunk_start = volume * chunk / num_chunks;
+      int64_t chunk_end = volume * (chunk + 1) / num_chunks;
+
+      engine.AddWork([=, &tmp_sets, &lock](int thread) {
+        OutLabel prev = old_bg_label;
+        OutLabel remapped = old_bg_label;
+        for (int64_t i = chunk_start; i < chunk_end; i++) {
+          OutLabel curr = labels[i];
+          if (curr != old_bg_label) {
+            if (curr != prev) {
+              {
+                std::lock_guard<spinlock> g(lock);
+                prev = curr;
+                // look up `ds` only when the value changes - this saves a lot of lookups
+                remapped = ds.find(labels, i);
+              }
+              tmp_sets[thread].insert(remapped);
+            } else {
+              labels[i] = remapped;
```

Comment:
I may be a bit lost, but either we don't need that else, or we should also `labels[i] = remapped` in the if as well. Otherwise, we leave it's value as a `prev`, and `curr = labels[i]` is equal to `prev`, isn't it?

Reply:
`find` sets it in the `if` branch.